Sources/OpenAPIRuntime/Conversion/MIMEType.swift

```diff
@@ -0,0 +1,156 @@
+    ///   - parameters: Any optional parameters.
+    public init(type: Token, subtype: Token, parameters: [String: String] = [:]) {
+        self.type = type
+        self.subtype = subtype
+        self.parameters = parameters
+    }
+
+    public static func == (lhs: MIMEType, rhs: MIMEType) -> Bool {
+        guard lhs.type == rhs.type else {
+            return false
+        }
+        guard lhs.subtype == rhs.subtype else {
+            return false
+        }
+        // Parameter names are case-insensitive, parameter values are
+        // case-sensitive.
+        guard lhs.parameters.count == rhs.parameters.count else {
+            return false
+        }
+        func normalizeKeyValue(key: String, value: String) -> (String, String) {
+            (key.lowercased(), value)
+        }
+        let normalizedLeftParams = Dictionary(
+            uniqueKeysWithValues: lhs.parameters.map(normalizeKeyValue)
+        )
+        let normalizedRightParams = Dictionary(
+            uniqueKeysWithValues: rhs.parameters.map(normalizeKeyValue)
+        )
+        return normalizedLeftParams == normalizedRightParams
+    }
+}
+
+extension MIMEType.Token: LosslessStringConvertible {
+    public init?(_ description: String) {
+        if description == "*" {
+            self = .wildcard
+        } else {
+            self = .concrete(description)
+        }
+    }
+
+    public var description: String {
+        switch self {
+        case .wildcard:
+            return "*"
+        case .concrete(let string):
+            return string
+        }
+    }
+}
+
+extension MIMEType: LosslessStringConvertible {
+    public init?(_ description: String) {
+        var components =
+            description
+            // Split by semicolon
+            .split(separator: ";")
+            .map(String.init)
+            // Trim leading/trailing spaces
+            .map { $0.soar_trimmingLeadingAndTrailingSpaces }
+        guard !components.isEmpty else {
+            return nil
+        }
+        let firstComponent = components.removeFirst()
+        let typeAndSubtype =
+            firstComponent
+            .split(separator: "/")
+            .map(String.init)
+            .compactMap(MIMEType.Token.init)
+        guard typeAndSubtype.count == 2 else {
+            return nil
+        }
+        func parseParameter(_ string: String) -> (String, String)? {
+            let components =
+                string
+                .split(separator: "=")
+                .map(String.init)
+            guard components.count == 2 else {
+                return nil
+            }
+            return (components[0], components[1])
+        }
+        let parameters =
+            components
+            .compactMap(parseParameter)
+        self.init(
+            type: typeAndSubtype[0],
+            subtype: typeAndSubtype[1],
+            parameters: Dictionary(
+                parameters,
+                uniquingKeysWith: { a, _ in a }
+            )
```

Comment:
Are we being overly permissive here? I skimmed the RFC and it the only thing I can interpret requirements about uniqueness of parameters is this:

> After the media type and subtype names, the remainder
   of the header field is simply a set of parameters, specified in an
   attribute=value notation.  The ordering of parameters is not
   significant.

The use of insignificant order and term "set", I infer uniqueness. Should we be rejecting MIMETypes with duplicates as it can only lead to confusion.

Additionally it makes this not LosslessStringConvertible because a round trip may lose some duplicate parameters.

I think we should be returning `nil` here, WDYT?

Reply:
> Additionally it makes this not LosslessStringConvertible because a round trip may lose some duplicate parameters.

That's a great point, but unfortunately there isn't just a `StringConvertible` protocol, which is what I was trying to do here. Should I define our own protocol to avoid the confusion with `Lossless`?

> I think we should be returning nil here, WDYT?

Totally reasonable, I read some advice on this and servers apparently usually keep the first, or the last value, when multiple are found. We can be stricter than that and completely reject (aka ignore) the header if there are multiple, but on the other hand, we don't ever _use_ the parameters, so we don't really care if parameters are duplicated, we only use the type/subtype. Maybe it's more philosophical around inputs in general, whether we should be strict even about inputs we don't use. I'm happy to go either way, with the above, do you still think we should discard the MIME type if it has multiple params with the same name?

Reply:
OK, if you feel like this is conventional then that's fine. I just went to the spec. Maybe it's worth dropping a comment here?